src/opentracing/Tracer.ts

```diff
@@ -0,0 +1,143 @@
+import { Tracer, Reference, Tags } from 'opentracing';
+import * as opentracing from 'opentracing';
+import ThundraSpan, { SpanEvent } from './Span';
+import ThundraSpanContext from './SpanContext';
+import ThundraRecorder from './Recorder';
+import ThundraSampler from './Sampler';
+import Utils from '../plugins/Utils';
+
+class ThundraTracer extends Tracer {
+  tags: any;
+  recorder: ThundraRecorder;
+  sampler: ThundraSampler;
+  activeSpans: Map<string, ThundraSpan>;
+
+  constructor(config: any = {}) {
+    super();
+
+    this.tags = config.tags;
+    this.recorder = config.recorder ? config.recorder : new ThundraRecorder();
+    this.sampler = config.sampler ? config.sampler : new ThundraSampler(1);
+    this.activeSpans = new Map<string, ThundraSpan>();
+  }
+
+  getActiveSpan(): ThundraSpan {
+    return this.recorder.getActiveSpan() ? this.recorder.getActiveSpan().value : null;
+  }
+
+  finishSpan(): void {
+    if (this.getActiveSpan()) {
+       this.getActiveSpan().finish();
+    }
+  }
+
+  getRecorder(): ThundraRecorder {
+    return this.recorder;
+  }
+
+  destroy() {
+    this.recorder.destroy();
+    this.activeSpans = null;
+  }
+
+  wrapper<T extends (...args: any[]) => any>(spanName: string, func: T): T {
+    // tslint:disable-next-line:no-angle-bracket-type-assertion
+    return <T> ((...args: any[]) => {
+      try {
+        this.startSpan(spanName);
+        const returnValue = func(...args);
+        this.finishSpan();
+        return returnValue;
+      } catch (error) {
+        this.finishSpan();
+      }
+    });
+  }
+
+  _startSpan(name: any, fields: any) {
+    const tags: any = {};
+    let span;
+    const rootTraceId = fields.rootTraceId ? fields.rootTraceId : Utils.generateId();
+
+    const parentContext = getParent(fields.references);
+
+    if (parentContext && !this.activeSpans.get(parentContext.spanId)) {
+        throw new Error('Invalid spanId : ' + parentContext.spanId);
+    }
+
+    if (fields) {
+      span = new ThundraSpan(this, {
+        operationName: fields.operationName || name,
+        parent: getParent(fields.references),
+        tags: Object.assign(tags, this.tags, fields.tags),
+        startTime: fields.startTime,
+        rootTraceId,
+        className: fields.className,
+        domainName: fields.domainName,
+      });
+    } else {
+      span = new ThundraSpan(this, {
+        operationName: name,
+        parent: this.getActiveSpan(),
```

Comment:
`parent` field above in the `if` statement is `ThundraSpanContext ` typed but as far as I see, in here, it is `ThundraSpan` typed (`ThundraSpan` typed `value` field of `SpanTreeNode`). 

So seems that at different execution paths, we pass different typed values for `parent` field to `ThundraSpan`

Reply:
Yes, you are right it should be 'this.getActiveSpan().spanContext'. It did not cause any problem, because we are tracking parent/child relation in the recorder. In our current structure, we do not use parent context.  This field will be beneficial when user sends data with custom recorder.